// @tsparticles/updater-twinkle/tsparticles.updater.twinkle.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Updater v4.1.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.updaters = global.__tsParticlesInternals.updaters || {}, global.__tsParticlesInternals.updaters.twinkle = global.__tsParticlesInternals.updaters.twinkle || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class TwinkleLinksValues {
        color;
        enable;
        frequency;
        opacity;
        constructor() {
            this.enable = false;
            this.frequency = 0.05;
            this.opacity = 1;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.color !== undefined) {
                this.color = engine.OptionsColor.create(this.color, data.color);
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.frequency !== undefined) {
                this.frequency = data.frequency;
            }
            if (data.opacity !== undefined) {
                this.opacity = engine.setRangeValue(data.opacity);
            }
        }
    }

    class TwinkleParticlesValues {
        enable;
        fillColor;
        frequency;
        opacity;
        strokeColor;
        constructor() {
            this.enable = false;
            this.frequency = 0.05;
            this.opacity = 1;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.fillColor !== undefined) {
                this.fillColor = engine.OptionsColor.create(this.fillColor, data.fillColor);
            }
            if (data.strokeColor !== undefined) {
                this.strokeColor = engine.OptionsColor.create(this.strokeColor, data.strokeColor);
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.frequency !== undefined) {
                this.frequency = data.frequency;
            }
            if (data.opacity !== undefined) {
                this.opacity = engine.setRangeValue(data.opacity);
            }
        }
    }

    class Twinkle {
        links;
        particles;
        constructor() {
            this.links = new TwinkleLinksValues();
            this.particles = new TwinkleParticlesValues();
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            this.links.load(data.links);
            this.particles.load(data.particles);
        }
    }

    class TwinkleUpdater {
        #container;
        #pluginManager;
        constructor(pluginManager, container) {
            this.#pluginManager = pluginManager;
            this.#container = container;
        }
        getColorStyles(particle, _context, _radius, opacity) {
            const pOptions = particle.options, container = this.#container, twinkleOptions = pOptions["twinkle"];
            if (!twinkleOptions) {
                return {};
            }
            const twinkle = twinkleOptions.particles, twinkling = twinkle.enable && engine.getRandom() < twinkle.frequency, zIndexOptions = particle.options.zIndex, zOffset = 1, zOpacityFactor = (zOffset - particle.zIndexFactor) ** zIndexOptions.opacityRate, twinklingOpacity = twinkling ? engine.getRangeValue(twinkle.opacity) * zOpacityFactor : opacity, twinkleFillRgb = engine.rangeColorToHsl(this.#pluginManager, twinkle.fillColor), twinkleStrokeRgb = engine.rangeColorToHsl(this.#pluginManager, twinkle.strokeColor), getTwinkleFillStyle = () => {
                if (!twinkleFillRgb) {
                    return undefined;
                }
                return engine.getStyleFromHsl(twinkleFillRgb, container.hdr, twinklingOpacity);
            }, getTwinkleStrokeStyle = () => {
                if (!twinkleStrokeRgb) {
                    return undefined;
                }
                return engine.getStyleFromHsl(twinkleStrokeRgb, container.hdr, twinklingOpacity);
            }, twinkleFillStyle = getTwinkleFillStyle(), twinkleStrokeStyle = getTwinkleStrokeStyle(), res = {}, needsTwinkle = twinkling && (!!twinkleFillStyle || !!twinkleStrokeStyle);
            res.fill = needsTwinkle ? twinkleFillStyle : undefined;
            res.stroke = needsTwinkle ? twinkleStrokeStyle : undefined;
            return res;
        }
        init() {
        }
        isEnabled(particle) {
            const pOptions = particle.options, twinkleOptions = pOptions.twinkle;
            if (!twinkleOptions) {
                return false;
            }
            return twinkleOptions.particles.enable;
        }
        loadOptions(options, ...sources) {
            options.twinkle ??= new Twinkle();
            for (const source of sources) {
                options.twinkle.load(source?.twinkle);
            }
        }
        update() {
        }
    }

    async function loadTwinkleUpdater(engine) {
        engine.checkVersion("4.1.1");
        await engine.pluginManager.register(e => {
            e.pluginManager.addParticleUpdater("twinkle", container => {
                return Promise.resolve(new TwinkleUpdater(e.pluginManager, container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadTwinkleUpdater = loadTwinkleUpdater;

    exports.loadTwinkleUpdater = loadTwinkleUpdater;

}));
Object.assign(globalThis.window || globalThis, { loadTwinkleUpdater: (globalThis.__tsParticlesInternals.updaters.twinkle || {}).loadTwinkleUpdater });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
