// @tsparticles/plugin-absorbers/esm/AbsorbersInteractor.js
import { ExternalInteractorBase, } from "@tsparticles/plugin-interactivity";
import { getDistance, isArray, isInArray, itemFromArray, } from "@tsparticles/engine";
import { Absorber } from "./Options/Classes/Absorber.js";
const absorbersMode = "absorbers";
export class AbsorbersInteractor extends ExternalInteractorBase {
    handleClickMode;
    maxDistance;
    #dragging = false;
    #draggingAbsorber;
    #instancesManager;
    constructor(container, instancesManager) {
        super(container);
        this.maxDistance = 0;
        this.#instancesManager = instancesManager;
        this.#instancesManager.initContainer(container);
        this.handleClickMode = (mode, interactivityData) => {
            const container = this.container, options = container.actualOptions, absorbers = options.interactivity.modes.absorbers;
            if (!absorbers || mode !== absorbersMode) {
                return;
            }
            const { clickPosition } = interactivityData.mouse;
            if (clickPosition) {
                const existingAbsorber = instancesManager
                    .getArray(this.container)
                    .some(t => getDistance(t.position, clickPosition) < t.size);
                if (existingAbsorber) {
                    return;
                }
            }
            const absorbersModeOptions = itemFromArray(absorbers) ?? new Absorber();
            void this.#instancesManager.addAbsorber(container, absorbersModeOptions, clickPosition);
        };
    }
    clear() {
    }
    init() {
    }
    interact(interactivityData, delta) {
        for (const particle of this.container.particles.filter(p => this.isEnabled(interactivityData, p))) {
            for (const absorber of this.#instancesManager.getArray(this.container)) {
                if (absorber.options.draggable) {
                    const mouse = interactivityData.mouse;
                    if (mouse.clicking && mouse.downPosition) {
                        const mouseDist = getDistance(absorber.position, mouse.downPosition);
                        if (mouseDist <= absorber.size) {
                            this.#dragging = true;
                            this.#draggingAbsorber = absorber;
                        }
                    }
                    else {
                        this.#dragging = false;
                        this.#draggingAbsorber = undefined;
                    }
                    if (this.#dragging && this.#draggingAbsorber == absorber && mouse.position) {
                        absorber.position.x = mouse.position.x;
                        absorber.position.y = mouse.position.y;
                    }
                }
                absorber.attract(particle, delta);
                if (particle.destroyed) {
                    break;
                }
            }
        }
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity).events;
        if (!mouse.clickPosition || !events.onClick.enable) {
            return false;
        }
        return isInArray(absorbersMode, events.onClick.mode);
    }
    loadModeOptions(options, ...sources) {
        options.absorbers ??= [];
        for (const source of sources) {
            if (!source) {
                continue;
            }
            if (isArray(source.absorbers)) {
                for (const absorber of source.absorbers) {
                    const tmp = new Absorber();
                    tmp.load(absorber);
                    options.absorbers.push(tmp);
                }
            }
            else {
                const tmp = new Absorber();
                tmp.load(source.absorbers);
                options.absorbers.push(tmp);
            }
        }
    }
    reset() {
    }
}
