// @tsparticles/plugin-absorbers/esm/AbsorbersPluginInstance.js
import { executeOnSingleOrMultiple, isArray, } from "@tsparticles/engine";
export class AbsorbersPluginInstance {
    #container;
    #instancesManager;
    constructor(container, instancesManager) {
        this.#container = container;
        this.#instancesManager = instancesManager;
        this.#instancesManager.initContainer(container);
    }
    draw(context) {
        for (const absorber of this.#instancesManager.getArray(this.#container)) {
            absorber.draw(context);
        }
    }
    async init() {
        const absorbers = this.#container.actualOptions.absorbers, promises = executeOnSingleOrMultiple(absorbers, async (absorber) => {
            await this.#instancesManager.addAbsorber(this.#container, absorber);
        });
        if (isArray(promises)) {
            await Promise.all(promises);
        }
        else {
            await promises;
        }
    }
    particleUpdate(particle, delta) {
        for (const absorber of this.#instancesManager.getArray(this.#container)) {
            absorber.attract(particle, delta);
            if (particle.destroyed) {
                break;
            }
        }
    }
    resize() {
        for (const absorber of this.#instancesManager.getArray(this.#container)) {
            absorber.resize();
        }
    }
    stop() {
        this.#instancesManager.clear(this.#container);
    }
}
