// @tsparticles/updater-wobble/tsparticles.updater.wobble.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Updater v4.1.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.updaters = global.__tsParticlesInternals.updaters || {}, global.__tsParticlesInternals.updaters.wobble = global.__tsParticlesInternals.updaters.wobble || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class WobbleSpeed {
        angle;
        move;
        constructor() {
            this.angle = 50;
            this.move = 10;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.angle !== undefined) {
                this.angle = engine.setRangeValue(data.angle);
            }
            if (data.move !== undefined) {
                this.move = engine.setRangeValue(data.move);
            }
        }
    }

    class Wobble {
        distance;
        enable;
        speed;
        constructor() {
            this.distance = 5;
            this.enable = false;
            this.speed = new WobbleSpeed();
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.distance !== undefined) {
                this.distance = engine.setRangeValue(data.distance);
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.speed !== undefined) {
                if (engine.isNumber(data.speed)) {
                    this.speed.load({ angle: data.speed });
                }
                else {
                    const rangeSpeed = data.speed;
                    if ("min" in rangeSpeed) {
                        this.speed.load({ angle: rangeSpeed });
                    }
                    else {
                        this.speed.load(data.speed);
                    }
                }
            }
        }
    }

    const defaultDistance$1 = 0, distanceFactor = 60;
    function updateWobble(container, particle, delta) {
        const { wobble: wobbleOptions } = particle.options, { wobble } = particle;
        if (!wobbleOptions?.enable || !wobble) {
            return;
        }
        const reduceFactor = container.retina.reduceFactor, angleSpeed = wobble.angleSpeed * delta.factor * reduceFactor, moveSpeed = wobble.moveSpeed * delta.factor * reduceFactor, distance = (moveSpeed * (particle.retina.wobbleDistance ?? defaultDistance$1)) / (engine.millisecondsToSeconds / distanceFactor), max = engine.doublePI, { position } = particle;
        wobble.angle += angleSpeed;
        if (wobble.angle > max) {
            wobble.angle -= max;
        }
        position.x += distance * Math.cos(wobble.angle);
        position.y += distance * Math.abs(Math.sin(wobble.angle));
    }

    const maxAngle = 360, moveSpeedFactor = 10, defaultDistance = 0;
    class WobbleUpdater {
        #container;
        constructor(container) {
            this.#container = container;
        }
        init(particle) {
            const wobbleOpt = particle.options.wobble;
            if (wobbleOpt?.enable) {
                particle.wobble = {
                    angle: engine.getRandom() * engine.doublePI,
                    angleSpeed: engine.getRangeValue(wobbleOpt.speed.angle) / maxAngle,
                    moveSpeed: engine.getRangeValue(wobbleOpt.speed.move) / moveSpeedFactor,
                };
            }
            else {
                particle.wobble = {
                    angle: 0,
                    angleSpeed: 0,
                    moveSpeed: 0,
                };
            }
            particle.retina.wobbleDistance =
                engine.getRangeValue(wobbleOpt?.distance ?? defaultDistance) * this.#container.retina.pixelRatio;
        }
        isEnabled(particle) {
            return !particle.destroyed && !particle.spawning && !!particle.options.wobble?.enable;
        }
        loadOptions(options, ...sources) {
            options.wobble ??= new Wobble();
            for (const source of sources) {
                options.wobble.load(source?.wobble);
            }
        }
        update(particle, delta) {
            if (!this.isEnabled(particle)) {
                return;
            }
            updateWobble(this.#container, particle, delta);
        }
    }

    async function loadWobbleUpdater(engine) {
        engine.checkVersion("4.1.1");
        await engine.pluginManager.register(e => {
            e.pluginManager.addParticleUpdater("wobble", container => {
                return Promise.resolve(new WobbleUpdater(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadWobbleUpdater = loadWobbleUpdater;

    exports.loadWobbleUpdater = loadWobbleUpdater;

}));
Object.assign(globalThis.window || globalThis, { loadWobbleUpdater: (globalThis.__tsParticlesInternals.updaters.wobble || {}).loadWobbleUpdater });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
