// @tsparticles/plugin-absorbers/tsparticles.plugin.absorbers.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Plugin v4.1.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.plugins = global.__tsParticlesInternals.plugins || {}, global.__tsParticlesInternals.plugins.absorbers = global.__tsParticlesInternals.plugins.absorbers || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    class AbsorberLife {
        count;
        delay;
        duration;
        wait;
        constructor() {
            this.wait = false;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.count !== undefined) {
                this.count = data.count;
            }
            if (data.delay !== undefined) {
                this.delay = engine.setRangeValue(data.delay);
            }
            if (data.duration !== undefined) {
                this.duration = engine.setRangeValue(data.duration);
            }
            if (data.wait !== undefined) {
                this.wait = data.wait;
            }
        }
    }

    class AbsorberSizeLimit {
        mass;
        radius;
        constructor() {
            this.radius = 0;
            this.mass = 0;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.mass !== undefined) {
                this.mass = data.mass;
            }
            if (data.radius !== undefined) {
                this.radius = data.radius;
            }
        }
    }

    class AbsorberSize extends engine.ValueWithRandom {
        density;
        limit;
        constructor() {
            super();
            this.density = 5;
            this.value = 50;
            this.limit = new AbsorberSizeLimit();
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            super.load(data);
            if (data.density !== undefined) {
                this.density = data.density;
            }
            if (engine.isNumber(data.limit)) {
                this.limit.radius = data.limit;
            }
            else {
                this.limit.load(data.limit);
            }
        }
    }

    class Absorber {
        color;
        destroy;
        draggable;
        life;
        name;
        opacity;
        orbits;
        position;
        size;
        constructor() {
            this.color = new engine.OptionsColor();
            this.color.value = "#000000";
            this.draggable = false;
            this.opacity = 1;
            this.destroy = true;
            this.orbits = false;
            this.life = new AbsorberLife();
            this.size = new AbsorberSize();
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.color !== undefined) {
                this.color = engine.OptionsColor.create(this.color, data.color);
            }
            if (data.draggable !== undefined) {
                this.draggable = data.draggable;
            }
            if (data.life !== undefined) {
                this.life.load(data.life);
            }
            this.name = data.name;
            if (data.opacity !== undefined) {
                this.opacity = data.opacity;
            }
            if (data.position !== undefined) {
                this.position = {};
                if (data.position.x !== undefined) {
                    this.position.x = engine.setRangeValue(data.position.x);
                }
                if (data.position.y !== undefined) {
                    this.position.y = engine.setRangeValue(data.position.y);
                }
            }
            if (data.size !== undefined) {
                this.size.load(data.size);
            }
            if (data.destroy !== undefined) {
                this.destroy = data.destroy;
            }
            if (data.orbits !== undefined) {
                this.orbits = data.orbits;
            }
        }
    }

    const absorbersMode = "absorbers";
    class AbsorbersInteractor extends pluginInteractivity.ExternalInteractorBase {
        handleClickMode;
        maxDistance;
        #dragging = false;
        #draggingAbsorber;
        #instancesManager;
        constructor(container, instancesManager) {
            super(container);
            this.maxDistance = 0;
            this.#instancesManager = instancesManager;
            this.#instancesManager.initContainer(container);
            this.handleClickMode = (mode, interactivityData) => {
                const container = this.container, options = container.actualOptions, absorbers = options.interactivity.modes.absorbers;
                if (!absorbers || mode !== absorbersMode) {
                    return;
                }
                const { clickPosition } = interactivityData.mouse;
                if (clickPosition) {
                    const existingAbsorber = instancesManager
                        .getArray(this.container)
                        .some(t => engine.getDistance(t.position, clickPosition) < t.size);
                    if (existingAbsorber) {
                        return;
                    }
                }
                const absorbersModeOptions = engine.itemFromArray(absorbers) ?? new Absorber();
                void this.#instancesManager.addAbsorber(container, absorbersModeOptions, clickPosition);
            };
        }
        clear() {
        }
        init() {
        }
        interact(interactivityData, delta) {
            for (const particle of this.container.particles.filter(p => this.isEnabled(interactivityData, p))) {
                for (const absorber of this.#instancesManager.getArray(this.container)) {
                    if (absorber.options.draggable) {
                        const mouse = interactivityData.mouse;
                        if (mouse.clicking && mouse.downPosition) {
                            const mouseDist = engine.getDistance(absorber.position, mouse.downPosition);
                            if (mouseDist <= absorber.size) {
                                this.#dragging = true;
                                this.#draggingAbsorber = absorber;
                            }
                        }
                        else {
                            this.#dragging = false;
                            this.#draggingAbsorber = undefined;
                        }
                        if (this.#dragging && this.#draggingAbsorber == absorber && mouse.position) {
                            absorber.position.x = mouse.position.x;
                            absorber.position.y = mouse.position.y;
                        }
                    }
                    absorber.attract(particle, delta);
                    if (particle.destroyed) {
                        break;
                    }
                }
            }
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity).events;
            if (!mouse.clickPosition || !events.onClick.enable) {
                return false;
            }
            return engine.isInArray(absorbersMode, events.onClick.mode);
        }
        loadModeOptions(options, ...sources) {
            options.absorbers ??= [];
            for (const source of sources) {
                if (!source) {
                    continue;
                }
                if (engine.isArray(source.absorbers)) {
                    for (const absorber of source.absorbers) {
                        const tmp = new Absorber();
                        tmp.load(absorber);
                        options.absorbers.push(tmp);
                    }
                }
                else {
                    const tmp = new Absorber();
                    tmp.load(source.absorbers);
                    options.absorbers.push(tmp);
                }
            }
        }
        reset() {
        }
    }

    const instancesManagers = new WeakMap();
    function getAbsorbersInstancesManager(e) {
        const pluginManager = e.pluginManager;
        let manager = instancesManagers.get(pluginManager);
        if (!manager) {
            manager = Promise.resolve().then(function () { return AbsorbersInstancesManager$1; })
                .then(({ AbsorbersInstancesManager }) => new AbsorbersInstancesManager(pluginManager))
                .catch((error) => {
                instancesManagers.delete(pluginManager);
                throw error;
            });
            instancesManagers.set(pluginManager, manager);
        }
        return manager;
    }

    async function loadAbsorbersInteraction(engine) {
        engine.checkVersion("4.1.1");
        await engine.pluginManager.register(async (e) => {
            const pluginManager = e.pluginManager, instancesManager = await getAbsorbersInstancesManager(e);
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            pluginManager.addInteractor?.("externalAbsorbers", container => {
                return Promise.resolve(new AbsorbersInteractor(container, instancesManager));
            });
        });
    }

    class AbsorbersPlugin {
        id = "absorbers";
        #instancesManager;
        constructor(instancesManager) {
            this.#instancesManager = instancesManager;
        }
        async getPlugin(container) {
            const { AbsorbersPluginInstance } = await Promise.resolve().then(function () { return AbsorbersPluginInstance$1; });
            return new AbsorbersPluginInstance(container, this.#instancesManager);
        }
        loadOptions(_container, options, source) {
            if (!this.needsPlugin(options) && !this.needsPlugin(source)) {
                return;
            }
            if (source?.absorbers) {
                options.absorbers = engine.executeOnSingleOrMultiple(source.absorbers, absorber => {
                    const tmp = new Absorber();
                    tmp.load(absorber);
                    return tmp;
                });
            }
        }
        needsPlugin(options) {
            if (!options) {
                return false;
            }
            const absorbers = options.absorbers;
            if (engine.isArray(absorbers)) {
                return !!absorbers.length;
            }
            else if (absorbers) {
                return true;
            }
            return false;
        }
    }

    async function loadAbsorbersPluginSimple(engine) {
        engine.checkVersion("4.1.1");
        await engine.pluginManager.register(async (e) => {
            const pluginManager = e.pluginManager, instancesManager = await getAbsorbersInstancesManager(e);
            pluginManager.addPlugin(new AbsorbersPlugin(instancesManager));
        });
    }

    async function loadAbsorbersPlugin(engine) {
        await loadAbsorbersPluginSimple(engine);
        await loadAbsorbersInteraction(engine);
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadAbsorbersPlugin = loadAbsorbersPlugin;

    const defaultIndex = 0;
    class AbsorbersInstancesManager {
        #containerArrays;
        #pluginManager;
        constructor(pluginManager) {
            this.#pluginManager = pluginManager;
            this.#containerArrays = new Map();
        }
        async addAbsorber(container, options, position) {
            const { AbsorberInstance } = await Promise.resolve().then(function () { return AbsorberInstance$1; }), absorber = new AbsorberInstance(this.#pluginManager, container, options, position), array = this.getArray(container);
            array.push(absorber);
            return absorber;
        }
        clear(container) {
            this.initContainer(container);
            this.#containerArrays.set(container, []);
        }
        getArray(container) {
            this.initContainer(container);
            let array = this.#containerArrays.get(container);
            if (!array) {
                array = [];
                this.#containerArrays.set(container, array);
            }
            return array;
        }
        initContainer(container) {
            if (this.#containerArrays.has(container)) {
                return;
            }
            this.#containerArrays.set(container, []);
            container.getAbsorber ??= (idxOrName) => {
                const array = this.getArray(container);
                return idxOrName === undefined || engine.isNumber(idxOrName)
                    ? array[idxOrName ?? defaultIndex]
                    : array.find(t => t.name === idxOrName);
            };
            container.addAbsorber ??= (options, position) => {
                return this.addAbsorber(container, options, position);
            };
        }
        removeAbsorber(container, absorber) {
            const index = this.getArray(container).indexOf(absorber), deleteCount = 1;
            if (index >= defaultIndex) {
                this.getArray(container).splice(index, deleteCount);
            }
        }
    }

    var AbsorbersInstancesManager$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        AbsorbersInstancesManager: AbsorbersInstancesManager
    });

    class AbsorbersPluginInstance {
        #container;
        #instancesManager;
        constructor(container, instancesManager) {
            this.#container = container;
            this.#instancesManager = instancesManager;
            this.#instancesManager.initContainer(container);
        }
        draw(context) {
            for (const absorber of this.#instancesManager.getArray(this.#container)) {
                absorber.draw(context);
            }
        }
        async init() {
            const absorbers = this.#container.actualOptions.absorbers, promises = engine.executeOnSingleOrMultiple(absorbers, async (absorber) => {
                await this.#instancesManager.addAbsorber(this.#container, absorber);
            });
            if (engine.isArray(promises)) {
                await Promise.all(promises);
            }
            else {
                await promises;
            }
        }
        particleUpdate(particle, delta) {
            for (const absorber of this.#instancesManager.getArray(this.#container)) {
                absorber.attract(particle, delta);
                if (particle.destroyed) {
                    break;
                }
            }
        }
        resize() {
            for (const absorber of this.#instancesManager.getArray(this.#container)) {
                absorber.resize();
            }
        }
        stop() {
            this.#instancesManager.clear(this.#container);
        }
    }

    var AbsorbersPluginInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        AbsorbersPluginInstance: AbsorbersPluginInstance
    });

    const squareExp = 2, absorbFactor = 0.033, minOrbitLength = 0, minRadius = 0, minMass = 0, minAngle = 0, maxAngle = engine.doublePI, maxDegreeAngle = 360, angleIncrementFactor = engine.identity / maxDegreeAngle, minVelocity = 0, defaultLifeDelay = 0, minLifeCount = 0, defaultSpawnDelay = 0, defaultLifeCount = -1;
    class AbsorberInstance {
        color;
        limit;
        mass;
        name;
        opacity;
        options;
        position;
        size;
        #container;
        #currentDuration;
        #currentSpawnDelay;
        #duration;
        #firstSpawn;
        #immortal;
        #initialPosition;
        #lifeCount;
        #pluginManager;
        #spawnDelay;
        constructor(pluginManager, container, options, position) {
            this.#container = container;
            this.#pluginManager = pluginManager;
            this.#currentDuration = 0;
            this.#currentSpawnDelay = 0;
            this.#initialPosition = position ? engine.Vector.create(position.x, position.y) : undefined;
            if (options instanceof Absorber) {
                this.options = options;
            }
            else {
                this.options = new Absorber();
                this.options.load(options);
            }
            this.name = this.options.name;
            this.opacity = this.options.opacity;
            this.size = engine.getRangeValue(this.options.size.value) * container.retina.pixelRatio;
            this.mass = this.size * this.options.size.density * container.retina.reduceFactor;
            const limit = this.options.size.limit;
            this.limit = {
                radius: limit.radius * container.retina.pixelRatio * container.retina.reduceFactor,
                mass: limit.mass,
            };
            this.color = engine.rangeColorToRgb(this.#pluginManager, this.options.color) ?? {
                b: 0,
                g: 0,
                r: 0,
            };
            this.position = this.#initialPosition?.copy() ?? this.#calcPosition();
            this.#firstSpawn = !this.options.life.wait;
            this.#lifeCount = this.options.life.count ?? defaultLifeCount;
            this.#immortal = this.#lifeCount <= minLifeCount;
            this.#spawnDelay = container.retina.reduceFactor
                ? (engine.getRangeValue(this.options.life.delay ?? defaultLifeDelay) * engine.millisecondsToSeconds) /
                    container.retina.reduceFactor
                : Infinity;
        }
        attract(particle, delta) {
            const container = this.#container, options = this.options, pos = particle.getPosition(), { dx, dy, distance } = engine.getDistances(this.position, pos), v = engine.Vector.create(dx, dy);
            v.length = (this.mass / Math.pow(distance, squareExp)) * container.retina.reduceFactor;
            if (distance < this.size + particle.getRadius()) {
                const sizeFactor = particle.getRadius() * absorbFactor * container.retina.pixelRatio * delta.factor;
                if ((this.size > particle.getRadius() && distance < this.size - particle.getRadius()) ||
                    (particle.absorberOrbit !== undefined && particle.absorberOrbit.length < minOrbitLength)) {
                    if (options.destroy) {
                        particle.destroy();
                    }
                    else {
                        particle.needsNewPosition = true;
                        this.#updateParticlePosition(particle, delta, v);
                    }
                }
                else {
                    if (options.destroy) {
                        particle.size.value -= sizeFactor;
                    }
                    this.#updateParticlePosition(particle, delta, v);
                }
                if (this.limit.radius <= minRadius || this.size < this.limit.radius) {
                    this.size += sizeFactor;
                }
                if (this.limit.mass <= minMass || this.mass < this.limit.mass) {
                    this.mass += sizeFactor * this.options.size.density * container.retina.reduceFactor;
                }
            }
            else {
                this.#updateParticlePosition(particle, delta, v);
            }
        }
        draw(context) {
            context.translate(this.position.x, this.position.y);
            context.beginPath();
            context.arc(engine.originPoint.x, engine.originPoint.y, this.size, minAngle, maxAngle, false);
            context.closePath();
            context.fillStyle = engine.getStyleFromRgb(this.color, this.#container.hdr, this.opacity);
            context.fill();
        }
        resize() {
            const initialPosition = this.#initialPosition;
            this.position =
                initialPosition && engine.isPointInside(initialPosition, this.#container.canvas.size, engine.Vector.origin)
                    ? initialPosition
                    : this.#calcPosition();
        }
        update(delta) {
            if (this.#firstSpawn) {
                this.#firstSpawn = false;
                this.#currentSpawnDelay = this.#spawnDelay ?? defaultSpawnDelay;
            }
            if (this.#duration !== undefined) {
                this.#currentDuration += delta.value;
                if (this.#currentDuration >= this.#duration) {
                    if (!this.#immortal) {
                        this.#lifeCount--;
                    }
                    if (this.#lifeCount > minLifeCount || this.#immortal) {
                        this.position = this.#calcPosition();
                        this.#spawnDelay = this.#container.retina.reduceFactor
                            ? (engine.getRangeValue(this.options.life.delay ?? defaultLifeDelay) * engine.millisecondsToSeconds) /
                                this.#container.retina.reduceFactor
                            : Infinity;
                    }
                    this.#currentDuration -= this.#duration;
                    this.#duration = undefined;
                }
            }
            if (this.#spawnDelay !== undefined) {
                this.#currentSpawnDelay += delta.value;
                if (this.#currentSpawnDelay >= this.#spawnDelay) {
                    this.#play();
                    this.#currentSpawnDelay -= this.#spawnDelay;
                    this.#spawnDelay = undefined;
                }
            }
        }
        #calcPosition = () => {
            const exactPosition = engine.calcPositionOrRandomFromSizeRanged({
                size: this.#container.canvas.size,
                position: this.options.position,
            });
            return engine.Vector.create(exactPosition.x, exactPosition.y);
        };
        #play = () => {
            if (!((this.#lifeCount > minLifeCount || this.#immortal || !this.options.life.count) &&
                (this.#firstSpawn || this.#currentSpawnDelay >= (this.#spawnDelay ?? defaultSpawnDelay)))) {
                return;
            }
            if (this.#lifeCount > minLifeCount || this.#immortal) {
                this.#prepareToDie();
            }
        };
        #prepareToDie = () => {
            const duration = this.options.life.duration !== undefined ? engine.getRangeValue(this.options.life.duration) : undefined, minDuration = 0;
            if ((this.#lifeCount > minLifeCount || this.#immortal) && duration !== undefined && duration > minDuration) {
                this.#duration = duration * engine.millisecondsToSeconds;
            }
        };
        #updateParticlePosition = (particle, delta, v) => {
            if (particle.destroyed) {
                return;
            }
            const container = this.#container, canvasSize = container.canvas.size;
            if (particle.needsNewPosition) {
                const newPosition = engine.calcPositionOrRandomFromSize({ size: canvasSize });
                particle.position.setTo(newPosition);
                particle.velocity.setTo(particle.initialVelocity);
                particle.absorberOrbit = undefined;
                particle.needsNewPosition = false;
            }
            if (this.options.orbits) {
                if (particle.absorberOrbit === undefined) {
                    particle.absorberOrbit = engine.Vector.origin;
                    particle.absorberOrbit.length = engine.getDistance(particle.getPosition(), this.position);
                    particle.absorberOrbit.angle = engine.getRandom() * maxAngle;
                }
                if (particle.absorberOrbit.length <= this.size && !this.options.destroy) {
                    const minSize = Math.min(canvasSize.width, canvasSize.height), offset = 1, randomOffset = 0.1, randomFactor = 0.2;
                    particle.absorberOrbit.length = minSize * (offset + (engine.getRandom() * randomFactor - randomOffset));
                }
                particle.absorberOrbitDirection ??=
                    particle.velocity.x >= minVelocity ? engine.RotateDirection.clockwise : engine.RotateDirection.counterClockwise;
                const orbitRadius = particle.absorberOrbit.length, orbitAngle = particle.absorberOrbit.angle, orbitDirection = particle.absorberOrbitDirection;
                particle.velocity.setTo(engine.Vector.origin);
                const maxSize = particle.size.max, sizeFactor = particle.options.move.size ? particle.getRadius() / maxSize : engine.identity, deltaFactor = delta.factor || engine.identity, baseSpeed = particle.retina.moveSpeed, moveSpeed = baseSpeed * sizeFactor * deltaFactor * engine.half;
                particle.position.x = this.position.x + orbitRadius * Math.cos(orbitAngle);
                particle.position.y =
                    this.position.y +
                        orbitRadius * (orbitDirection === engine.RotateDirection.clockwise ? engine.identity : -engine.identity) * Math.sin(orbitAngle);
                particle.absorberOrbit.length = Math.max(minOrbitLength, particle.absorberOrbit.length - v.length);
                particle.absorberOrbit.angle += moveSpeed * angleIncrementFactor * container.retina.reduceFactor;
            }
            else {
                particle.velocity.addTo(v);
            }
        };
    }

    var AbsorberInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        AbsorberInstance: AbsorberInstance
    });

    exports.loadAbsorbersPlugin = loadAbsorbersPlugin;

}));
Object.assign(globalThis.window || globalThis, { loadAbsorbersPlugin: (globalThis.__tsParticlesInternals.plugins.absorbers || {}).loadAbsorbersPlugin });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
