// @tsparticles/plugin-emitters/esm/Options/Classes/Emitter.js
import { AnimatableColor, deepExtend, executeOnSingleOrMultiple, isNull, setRangeValue, } from "@tsparticles/engine";
import { EmitterLife } from "./EmitterLife.js";
import { EmitterRate } from "./EmitterRate.js";
import { EmitterShape } from "./EmitterShape.js";
import { EmitterSize } from "./EmitterSize.js";
import { EmitterSpawn } from "./EmitterSpawn.js";
export class Emitter {
    autoPlay;
    direction;
    domId;
    fill;
    life;
    name;
    particles;
    position;
    rate;
    shape;
    size;
    spawn;
    spawnFillColor;
    spawnStrokeColor;
    startCount;
    constructor() {
        this.autoPlay = true;
        this.fill = true;
        this.life = new EmitterLife();
        this.rate = new EmitterRate();
        this.shape = new EmitterShape();
        this.spawn = new EmitterSpawn();
        this.startCount = 0;
    }
    load(data) {
        if (isNull(data)) {
            return;
        }
        if (data.autoPlay !== undefined) {
            this.autoPlay = data.autoPlay;
        }
        if (data.size !== undefined) {
            this.size ??= new EmitterSize();
            this.size.load(data.size);
        }
        if (data.direction !== undefined) {
            this.direction = data.direction;
        }
        this.domId = data.domId;
        if (data.fill !== undefined) {
            this.fill = data.fill;
        }
        this.life.load(data.life);
        this.name = data.name;
        this.particles = executeOnSingleOrMultiple(data.particles, particles => {
            return deepExtend({}, particles);
        });
        this.rate.load(data.rate);
        this.shape.load(data.shape);
        this.spawn.load(data.spawn);
        if (data.position !== undefined) {
            this.position = {};
            if (data.position.x !== undefined) {
                this.position.x = setRangeValue(data.position.x);
            }
            if (data.position.y !== undefined) {
                this.position.y = setRangeValue(data.position.y);
            }
        }
        if (data.spawnFillColor !== undefined) {
            this.spawnFillColor ??= new AnimatableColor();
            this.spawnFillColor.load(data.spawnFillColor);
        }
        if (data.spawnStrokeColor !== undefined) {
            this.spawnStrokeColor ??= new AnimatableColor();
            this.spawnStrokeColor.load(data.spawnStrokeColor);
        }
        if (data.startCount !== undefined) {
            this.startCount = data.startCount;
        }
    }
}
