// @tsparticles/plugin-absorbers/browser/AbsorbersInstancesManager.js
import { isNumber } from "@tsparticles/engine";
const defaultIndex = 0;
export class AbsorbersInstancesManager {
    #containerArrays;
    #pluginManager;
    constructor(pluginManager) {
        this.#pluginManager = pluginManager;
        this.#containerArrays = new Map();
    }
    async addAbsorber(container, options, position) {
        const { AbsorberInstance } = await import("./AbsorberInstance.js"), absorber = new AbsorberInstance(this.#pluginManager, container, options, position), array = this.getArray(container);
        array.push(absorber);
        return absorber;
    }
    clear(container) {
        this.initContainer(container);
        this.#containerArrays.set(container, []);
    }
    getArray(container) {
        this.initContainer(container);
        let array = this.#containerArrays.get(container);
        if (!array) {
            array = [];
            this.#containerArrays.set(container, array);
        }
        return array;
    }
    initContainer(container) {
        if (this.#containerArrays.has(container)) {
            return;
        }
        this.#containerArrays.set(container, []);
        container.getAbsorber ??= (idxOrName) => {
            const array = this.getArray(container);
            return idxOrName === undefined || isNumber(idxOrName)
                ? array[idxOrName ?? defaultIndex]
                : array.find(t => t.name === idxOrName);
        };
        container.addAbsorber ??= (options, position) => {
            return this.addAbsorber(container, options, position);
        };
    }
    removeAbsorber(container, absorber) {
        const index = this.getArray(container).indexOf(absorber), deleteCount = 1;
        if (index >= defaultIndex) {
            this.getArray(container).splice(index, deleteCount);
        }
    }
}
