// @tsparticles/updater-destroy/tsparticles.updater.destroy.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Updater v4.1.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.updaters = global.__tsParticlesInternals.updaters || {}, global.__tsParticlesInternals.updaters.destroy = global.__tsParticlesInternals.updaters.destroy || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class DestroyBounds {
        bottom;
        left;
        right;
        top;
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.bottom !== undefined) {
                this.bottom = engine.setRangeValue(data.bottom);
            }
            if (data.left !== undefined) {
                this.left = engine.setRangeValue(data.left);
            }
            if (data.right !== undefined) {
                this.right = engine.setRangeValue(data.right);
            }
            if (data.top !== undefined) {
                this.top = engine.setRangeValue(data.top);
            }
        }
    }

    var DestroyMode;
    (function (DestroyMode) {
        DestroyMode["explode"] = "explode";
        DestroyMode["none"] = "none";
        DestroyMode["split"] = "split";
    })(DestroyMode || (DestroyMode = {}));

    class Explode {
        maxSizeFactor;
        speed;
        constructor() {
            this.maxSizeFactor = 3;
            this.speed = 2;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.maxSizeFactor !== undefined) {
                this.maxSizeFactor = data.maxSizeFactor;
            }
            if (data.speed !== undefined) {
                this.speed = data.speed;
            }
        }
    }

    class SplitFactor extends engine.ValueWithRandom {
        constructor() {
            super();
            this.value = 3;
        }
    }

    class SplitRate extends engine.ValueWithRandom {
        constructor() {
            super();
            this.value = { min: 4, max: 9 };
        }
    }

    class Split {
        count;
        factor;
        fillColor;
        fillColorOffset;
        particles;
        rate;
        sizeOffset;
        strokeColor;
        strokeColorOffset;
        constructor() {
            this.count = 1;
            this.factor = new SplitFactor();
            this.rate = new SplitRate();
            this.sizeOffset = true;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.fillColor !== undefined) {
                this.fillColor = engine.OptionsColor.create(this.fillColor, data.fillColor);
            }
            if (data.strokeColor !== undefined) {
                this.strokeColor = engine.OptionsColor.create(this.strokeColor, data.strokeColor);
            }
            if (data.count !== undefined) {
                this.count = data.count;
            }
            this.factor.load(data.factor);
            this.rate.load(data.rate);
            this.particles = engine.executeOnSingleOrMultiple(data.particles, particles => {
                return engine.deepExtend({}, particles);
            });
            if (data.sizeOffset !== undefined) {
                this.sizeOffset = data.sizeOffset;
            }
            if (data.fillColorOffset) {
                this.fillColorOffset = this.fillColorOffset ?? {};
                if (data.fillColorOffset.h !== undefined) {
                    this.fillColorOffset.h = data.fillColorOffset.h;
                }
                if (data.fillColorOffset.s !== undefined) {
                    this.fillColorOffset.s = data.fillColorOffset.s;
                }
                if (data.fillColorOffset.l !== undefined) {
                    this.fillColorOffset.l = data.fillColorOffset.l;
                }
            }
            if (data.strokeColorOffset) {
                this.strokeColorOffset = this.strokeColorOffset ?? {};
                if (data.strokeColorOffset.h !== undefined) {
                    this.strokeColorOffset.h = data.strokeColorOffset.h;
                }
                if (data.strokeColorOffset.s !== undefined) {
                    this.strokeColorOffset.s = data.strokeColorOffset.s;
                }
                if (data.strokeColorOffset.l !== undefined) {
                    this.strokeColorOffset.l = data.strokeColorOffset.l;
                }
            }
        }
    }

    class Destroy {
        bounds;
        explode;
        mode;
        split;
        constructor() {
            this.bounds = new DestroyBounds();
            this.explode = new Explode();
            this.mode = DestroyMode.none;
            this.split = new Split();
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.mode) {
                this.mode = data.mode;
            }
            if (data.bounds) {
                this.bounds.load(data.bounds);
            }
            this.explode.load(data.explode);
            this.split.load(data.split);
        }
    }

    const defaultOffset = 0, minDestroySize = 0.5, defaultSplitCount = 0, increment = 1, unbreakableTime = 500, minSplitCount = 0, minHue = 0, hueRange = 360, minSaturation = 0, maxSaturation = 100, minLightness = 0, maxLightness = 100;
    function createParentColor(parentColor) {
        return engine.AnimatableColor.create(undefined, {
            value: {
                hsl: parentColor,
            },
        });
    }
    function createOffsetColor(parentColor, offset) {
        const offsetH = engine.getRangeValue(offset.h ?? defaultOffset), offsetS = engine.getRangeValue(offset.s ?? defaultOffset), offsetL = engine.getRangeValue(offset.l ?? defaultOffset), h = (parentColor.h + offsetH) % hueRange, s = Math.max(minSaturation, Math.min(maxSaturation, parentColor.s + offsetS)), l = Math.max(minLightness, Math.min(maxLightness, parentColor.l + offsetL));
        return engine.AnimatableColor.create(undefined, {
            value: {
                hsl: {
                    h: h < minHue ? h + hueRange : h,
                    s,
                    l,
                },
            },
        });
    }
    function resolveSplitColor(offset, splitColor, splitParticlesColor, parentColor) {
        if (offset && parentColor) {
            return createOffsetColor(parentColor, offset);
        }
        if (splitColor) {
            return engine.AnimatableColor.create(undefined, splitColor);
        }
        if (splitParticlesColor) {
            return engine.AnimatableColor.create(undefined, splitParticlesColor);
        }
        return parentColor ? createParentColor(parentColor) : undefined;
    }
    function addSplitParticle(pluginManager, container, parent, splitParticlesOptions) {
        const destroyOptions = parent.options.destroy;
        if (!destroyOptions) {
            return;
        }
        const splitOptions = destroyOptions.split, splitParticleOptions = engine.loadParticlesOptions(pluginManager, container, parent.options), splitFillColor = engine.itemFromSingleOrMultiple(splitOptions.fillColor), splitStrokeColor = engine.itemFromSingleOrMultiple(splitOptions.strokeColor), parentFillColor = parent.getFillColor(), parentStrokeColor = parent.getStrokeColor();
        splitParticleOptions.move.load({
            center: {
                x: parent.position.x,
                y: parent.position.y,
                mode: engine.PixelMode.precise,
            },
        });
        const factor = engine.identity / engine.getRangeValue(splitOptions.factor.value), sizeOptions = splitParticleOptions["size"];
        if (sizeOptions) {
            if (engine.isNumber(sizeOptions.value)) {
                sizeOptions.value *= factor;
            }
            else {
                sizeOptions.value.min *= factor;
                sizeOptions.value.max *= factor;
            }
        }
        splitParticleOptions.load(splitParticlesOptions);
        const splitParticlePaintOptions = engine.itemFromSingleOrMultiple(splitParticleOptions.paint), splitParticleFillOptions = splitParticlePaintOptions?.fill, splitParticleStrokeOptions = splitParticlePaintOptions?.stroke, fillColor = resolveSplitColor(splitOptions.fillColorOffset, splitFillColor, splitParticleFillOptions?.color, parentFillColor), strokeColor = resolveSplitColor(splitOptions.strokeColorOffset, splitStrokeColor, splitParticleStrokeOptions?.color, parentStrokeColor);
        if (fillColor && splitParticleFillOptions) {
            splitParticleFillOptions.color = fillColor;
        }
        if (strokeColor && splitParticleStrokeOptions) {
            splitParticleStrokeOptions.color = strokeColor;
        }
        const offset = splitOptions.sizeOffset ? engine.setRangeValue(-parent.size.value, parent.size.value) : defaultOffset, position = {
            x: parent.position.x + engine.randomInRangeValue(offset),
            y: parent.position.y + engine.randomInRangeValue(offset),
        };
        return container.particles.addParticle(position, splitParticleOptions, parent.group, (particle) => {
            if (particle.size.value < minDestroySize) {
                return false;
            }
            particle.velocity.length = engine.randomInRangeValue(engine.setRangeValue(parent.velocity.length, particle.velocity.length));
            particle.splitCount = (parent.splitCount ?? defaultSplitCount) + increment;
            particle.unbreakable = true;
            particle.unbreakableUntil = performance.now() + unbreakableTime;
            return true;
        });
    }
    function split(pluginManager, container, particle) {
        const destroyOptions = particle.options.destroy;
        if (!destroyOptions) {
            return;
        }
        const splitOptions = destroyOptions.split;
        if (splitOptions.count >= minSplitCount) {
            if (particle.splitCount === undefined || particle.splitCount > splitOptions.count) {
                return;
            }
            particle.splitCount++;
        }
        const rate = engine.getRangeValue(splitOptions.rate.value), particlesSplitOptions = engine.itemFromSingleOrMultiple(splitOptions.particles);
        for (let i = 0; i < rate; i++) {
            addSplitParticle(pluginManager, container, particle, particlesSplitOptions);
        }
    }

    const defaultDeltaFactor = 1, fpsFactor = 60, minExplodeSpeed = 0.01, maxExplodeProgress = 1;
    class DestroyUpdater {
        #container;
        #pluginManager;
        constructor(pluginManager, container) {
            this.#container = container;
            this.#pluginManager = pluginManager;
        }
        init(particle) {
            const container = this.#container, particlesOptions = particle.options, destroyOptions = particlesOptions.destroy;
            if (!destroyOptions) {
                return;
            }
            particle.exploding = undefined;
            particle.splitCount = 0;
            const destroyBoundsOptions = destroyOptions.bounds;
            particle.destroyBounds ??= {};
            const { bottom, left, right, top } = destroyBoundsOptions, { destroyBounds } = particle, canvasSize = container.canvas.size;
            if (bottom) {
                destroyBounds.bottom = (engine.getRangeValue(bottom) * canvasSize.height) / engine.percentDenominator;
            }
            if (left) {
                destroyBounds.left = (engine.getRangeValue(left) * canvasSize.width) / engine.percentDenominator;
            }
            if (right) {
                destroyBounds.right = (engine.getRangeValue(right) * canvasSize.width) / engine.percentDenominator;
            }
            if (top) {
                destroyBounds.top = (engine.getRangeValue(top) * canvasSize.height) / engine.percentDenominator;
            }
        }
        isEnabled(particle) {
            const destroyParticle = particle;
            return !destroyParticle.destroyed || !!destroyParticle.exploding;
        }
        loadOptions(options, ...sources) {
            options.destroy ??= new Destroy();
            for (const source of sources) {
                options.destroy.load(source?.destroy);
            }
        }
        particleDestroyed(particle, override) {
            if (override) {
                return;
            }
            const destroyOptions = particle.options.destroy;
            switch (destroyOptions?.mode) {
                case DestroyMode.split:
                    split(this.#pluginManager, this.#container, particle);
                    break;
                case DestroyMode.explode: {
                    if (particle.exploding) {
                        particle.destroyed = false;
                        break;
                    }
                    const { explode } = destroyOptions, initialSize = particle.size.value, maxSize = initialSize * explode.maxSizeFactor, opacity = particle.getOpacity();
                    particle.exploding = {
                        initialFillOpacity: opacity.fillOpacity,
                        initialSize,
                        initialStrokeOpacity: opacity.strokeOpacity,
                        maxSize,
                        progress: 0,
                        speed: Math.max(explode.speed, minExplodeSpeed),
                    };
                    particle.fillOpacity = opacity.fillOpacity;
                    particle.strokeOpacity = opacity.strokeOpacity;
                    particle.destroyed = false;
                    break;
                }
            }
        }
        update(particle, delta) {
            if (particle.exploding) {
                const explosionState = particle.exploding, deltaFactor = delta.factor || defaultDeltaFactor;
                explosionState.progress = Math.min(maxExplodeProgress, explosionState.progress + (explosionState.speed * deltaFactor) / fpsFactor);
                const progress = explosionState.progress;
                particle.size.value =
                    explosionState.initialSize + (explosionState.maxSize - explosionState.initialSize) * progress;
                particle.fillOpacity = explosionState.initialFillOpacity * (maxExplodeProgress - progress);
                particle.strokeOpacity = explosionState.initialStrokeOpacity * (maxExplodeProgress - progress);
                if (progress >= maxExplodeProgress) {
                    particle.exploding = undefined;
                    particle.destroy(true);
                }
                return;
            }
            if (particle.unbreakableUntil !== undefined && performance.now() >= particle.unbreakableUntil) {
                particle.unbreakable = false;
                particle.unbreakableUntil = undefined;
            }
            const position = particle.getPosition(), bounds = particle.destroyBounds;
            if (!bounds) {
                return;
            }
            if ((bounds.bottom !== undefined && position.y >= bounds.bottom) ||
                (bounds.left !== undefined && position.x <= bounds.left) ||
                (bounds.right !== undefined && position.x >= bounds.right) ||
                (bounds.top !== undefined && position.y <= bounds.top)) {
                particle.destroy();
            }
        }
    }

    async function loadDestroyUpdater(engine) {
        engine.checkVersion("4.1.1");
        await engine.pluginManager.register(e => {
            e.pluginManager.addParticleUpdater("destroy", container => {
                return Promise.resolve(new DestroyUpdater(e.pluginManager, container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadDestroyUpdater = loadDestroyUpdater;

    exports.loadDestroyUpdater = loadDestroyUpdater;

}));
Object.assign(globalThis.window || globalThis, { loadDestroyUpdater: (globalThis.__tsParticlesInternals.updaters.destroy || {}).loadDestroyUpdater });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
