// @tsparticles/interaction-external-trail/tsparticles.interaction.external.trail.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* External Interaction v4.1.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.externalTrail = global.__tsParticlesInternals.interactions.externalTrail || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    class TrailColorWeight {
        x;
        y;
        constructor() {
            this.x = 0;
            this.y = 0;
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.x !== undefined) {
                this.x = data.x;
            }
            if (data.y !== undefined) {
                this.y = data.y;
            }
        }
    }

    class TrailColorComponent {
        value;
        weights;
        constructor() {
            this.weights = new TrailColorWeight();
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.weights) {
                this.weights ??= new TrailColorWeight();
                this.weights.load(data.weights);
            }
            if (!engine.isNull(data.value)) {
                this.value = engine.setRangeValue(data.value);
            }
        }
    }

    class TrailColorCoords {
        h;
        l;
        s;
        constructor() {
            this.h = new TrailColorComponent();
            this.s = new TrailColorComponent();
            this.l = new TrailColorComponent();
        }
        load(data) {
            if (!data) {
                return;
            }
            if (data.h) {
                this.h ??= new TrailColorComponent();
                this.h.load(data.h);
            }
            if (data.s) {
                this.s ??= new TrailColorComponent();
                this.s.load(data.s);
            }
            if (data.l) {
                this.l ??= new TrailColorComponent();
                this.l.load(data.l);
            }
        }
    }

    class Trail {
        colorCoords;
        delay;
        particles;
        pauseOnStop;
        quantity;
        constructor() {
            this.delay = 1;
            this.pauseOnStop = false;
            this.quantity = 1;
            this.colorCoords = new TrailColorCoords();
        }
        load(data) {
            if (engine.isNull(data))
                return;
            if (data.delay !== undefined) {
                this.delay = data.delay;
            }
            if (data.quantity !== undefined) {
                this.quantity = data.quantity;
            }
            if (data.pauseOnStop !== undefined) {
                this.pauseOnStop = data.pauseOnStop;
            }
            if (data.particles !== undefined) {
                this.particles = engine.deepExtend({}, data.particles);
            }
            if (data.colorCoords) {
                this.colorCoords ??= new TrailColorCoords();
                this.colorCoords.load(data.colorCoords);
            }
        }
    }

    const trailMode = "trail", defaultMin = 0, defaultWeight = 0;
    class TrailMaker extends pluginInteractivity.ExternalInteractorBase {
        maxDistance = 0;
        #delay;
        #lastPosition;
        #pluginManager;
        constructor(pluginManager, container) {
            super(container);
            this.#pluginManager = pluginManager;
            this.#delay = 0;
        }
        clear() {
        }
        init() {
        }
        interact(interactivityData, delta) {
            const container = this.container;
            if (!container.retina.reduceFactor)
                return;
            const options = container.actualOptions, trailOptions = options.interactivity?.modes.trail;
            if (!trailOptions)
                return;
            const optDelay = (trailOptions.delay * engine.millisecondsToSeconds) / this.container.retina.reduceFactor;
            if (this.#delay < optDelay) {
                this.#delay += delta.value;
            }
            if (this.#delay < optDelay)
                return;
            const mousePos = interactivityData.mouse.position, canEmit = !(trailOptions.pauseOnStop &&
                (mousePos === this.#lastPosition ||
                    (mousePos?.x === this.#lastPosition?.x && mousePos?.y === this.#lastPosition?.y)));
            if (mousePos) {
                this.#lastPosition = { ...mousePos };
            }
            else {
                this.#lastPosition = undefined;
            }
            if (canEmit && mousePos) {
                let particleOptions = trailOptions.particles;
                const colorCoords = trailOptions.colorCoords;
                if (colorCoords) {
                    const { width, height } = container.canvas.size, norm = {
                        x: mousePos.x / width,
                        y: mousePos.y / height,
                    }, hasWeights = (comp) => {
                        return !!(comp?.weights?.x ?? comp?.weights?.y);
                    }, calculateValue = (comp, originalValue, defaultMax) => {
                        if (!hasWeights(comp))
                            return undefined;
                        const w = comp?.weights, factor = norm.x * (w?.x ?? defaultWeight) + norm.y * (w?.y ?? defaultWeight), rangeSource = comp?.value ?? originalValue, min = engine.getRangeMin(rangeSource ?? defaultMin), max = engine.getRangeMax(rangeSource ?? defaultMax), result = min + factor * (max - min);
                        return Math.min(max, Math.max(min, result));
                    }, trailPaintOptions = trailOptions.particles?.paint
                        ? engine.itemFromSingleOrMultiple(trailOptions.particles.paint)
                        : undefined, fillData = trailPaintOptions?.fill, baseHsl = fillData
                        ? engine.rangeColorToHsl(this.#pluginManager, engine.AnimatableColor.create(undefined, fillData.color))
                        : undefined, h = calculateValue(colorCoords.h, baseHsl?.h, engine.hMax), s = calculateValue(colorCoords.s, baseHsl?.s, engine.sMax), l = calculateValue(colorCoords.l, baseHsl?.l, engine.lMax);
                    if (h !== undefined || s !== undefined || l !== undefined) {
                        particleOptions = engine.deepExtend({}, trailOptions.particles, {
                            paint: {
                                fill: {
                                    color: {
                                        value: {
                                            h: h ?? baseHsl?.h,
                                            s: s ?? baseHsl?.s,
                                            l: l ?? baseHsl?.l,
                                        },
                                    },
                                },
                            },
                        });
                    }
                }
                container.particles.push(trailOptions.quantity, mousePos, particleOptions);
            }
            this.#delay -= optDelay;
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity)?.events;
            return (!!events &&
                ((mouse.clicking && mouse.inside && !!mouse.position && engine.isInArray(trailMode, events.onClick.mode)) ||
                    (mouse.inside && !!mouse.position && engine.isInArray(trailMode, events.onHover.mode))));
        }
        loadModeOptions(options, ...sources) {
            options.trail ??= new Trail();
            for (const source of sources) {
                options.trail.load(source?.trail);
            }
        }
        reset() {
        }
    }

    async function loadExternalTrailInteraction(engine) {
        engine.checkVersion("4.1.1");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addInteractor?.("externalTrail", container => {
                return Promise.resolve(new TrailMaker(e.pluginManager, container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadExternalTrailInteraction = loadExternalTrailInteraction;

    exports.Trail = Trail;
    exports.loadExternalTrailInteraction = loadExternalTrailInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadExternalTrailInteraction: (globalThis.__tsParticlesInternals.interactions.externalTrail || {}).loadExternalTrailInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
