// @tsparticles/interaction-external-trail/cjs/TrailMaker.js
import { AnimatableColor, deepExtend, getRangeMax, getRangeMin, hMax, isInArray, itemFromSingleOrMultiple, lMax, millisecondsToSeconds, rangeColorToHsl, sMax, } from "@tsparticles/engine";
import { ExternalInteractorBase, } from "@tsparticles/plugin-interactivity";
import { Trail } from "./Options/Classes/Trail.js";
const trailMode = "trail", defaultMin = 0, defaultWeight = 0;
export class TrailMaker extends ExternalInteractorBase {
    maxDistance = 0;
    #delay;
    #lastPosition;
    #pluginManager;
    constructor(pluginManager, container) {
        super(container);
        this.#pluginManager = pluginManager;
        this.#delay = 0;
    }
    clear() {
    }
    init() {
    }
    interact(interactivityData, delta) {
        const container = this.container;
        if (!container.retina.reduceFactor)
            return;
        const options = container.actualOptions, trailOptions = options.interactivity?.modes.trail;
        if (!trailOptions)
            return;
        const optDelay = (trailOptions.delay * millisecondsToSeconds) / this.container.retina.reduceFactor;
        if (this.#delay < optDelay) {
            this.#delay += delta.value;
        }
        if (this.#delay < optDelay)
            return;
        const mousePos = interactivityData.mouse.position, canEmit = !(trailOptions.pauseOnStop &&
            (mousePos === this.#lastPosition ||
                (mousePos?.x === this.#lastPosition?.x && mousePos?.y === this.#lastPosition?.y)));
        if (mousePos) {
            this.#lastPosition = { ...mousePos };
        }
        else {
            this.#lastPosition = undefined;
        }
        if (canEmit && mousePos) {
            let particleOptions = trailOptions.particles;
            const colorCoords = trailOptions.colorCoords;
            if (colorCoords) {
                const { width, height } = container.canvas.size, norm = {
                    x: mousePos.x / width,
                    y: mousePos.y / height,
                }, hasWeights = (comp) => {
                    return !!(comp?.weights?.x ?? comp?.weights?.y);
                }, calculateValue = (comp, originalValue, defaultMax) => {
                    if (!hasWeights(comp))
                        return undefined;
                    const w = comp?.weights, factor = norm.x * (w?.x ?? defaultWeight) + norm.y * (w?.y ?? defaultWeight), rangeSource = comp?.value ?? originalValue, min = getRangeMin(rangeSource ?? defaultMin), max = getRangeMax(rangeSource ?? defaultMax), result = min + factor * (max - min);
                    return Math.min(max, Math.max(min, result));
                }, trailPaintOptions = trailOptions.particles?.paint
                    ? itemFromSingleOrMultiple(trailOptions.particles.paint)
                    : undefined, fillData = trailPaintOptions?.fill, baseHsl = fillData
                    ? rangeColorToHsl(this.#pluginManager, AnimatableColor.create(undefined, fillData.color))
                    : undefined, h = calculateValue(colorCoords.h, baseHsl?.h, hMax), s = calculateValue(colorCoords.s, baseHsl?.s, sMax), l = calculateValue(colorCoords.l, baseHsl?.l, lMax);
                if (h !== undefined || s !== undefined || l !== undefined) {
                    particleOptions = deepExtend({}, trailOptions.particles, {
                        paint: {
                            fill: {
                                color: {
                                    value: {
                                        h: h ?? baseHsl?.h,
                                        s: s ?? baseHsl?.s,
                                        l: l ?? baseHsl?.l,
                                    },
                                },
                            },
                        },
                    });
                }
            }
            container.particles.push(trailOptions.quantity, mousePos, particleOptions);
        }
        this.#delay -= optDelay;
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity)?.events;
        return (!!events &&
            ((mouse.clicking && mouse.inside && !!mouse.position && isInArray(trailMode, events.onClick.mode)) ||
                (mouse.inside && !!mouse.position && isInArray(trailMode, events.onHover.mode))));
    }
    loadModeOptions(options, ...sources) {
        options.trail ??= new Trail();
        for (const source of sources) {
            options.trail.load(source?.trail);
        }
    }
    reset() {
    }
}
