// @tsparticles/updater-twinkle/browser/TwinkleUpdater.js
import { getRandom, getRangeValue, getStyleFromHsl, rangeColorToHsl, } from "@tsparticles/engine";
import { Twinkle } from "./Options/Classes/Twinkle.js";
export class TwinkleUpdater {
    #container;
    #pluginManager;
    constructor(pluginManager, container) {
        this.#pluginManager = pluginManager;
        this.#container = container;
    }
    getColorStyles(particle, _context, _radius, opacity) {
        const pOptions = particle.options, container = this.#container, twinkleOptions = pOptions["twinkle"];
        if (!twinkleOptions) {
            return {};
        }
        const twinkle = twinkleOptions.particles, twinkling = twinkle.enable && getRandom() < twinkle.frequency, zIndexOptions = particle.options.zIndex, zOffset = 1, zOpacityFactor = (zOffset - particle.zIndexFactor) ** zIndexOptions.opacityRate, twinklingOpacity = twinkling ? getRangeValue(twinkle.opacity) * zOpacityFactor : opacity, twinkleFillRgb = rangeColorToHsl(this.#pluginManager, twinkle.fillColor), twinkleStrokeRgb = rangeColorToHsl(this.#pluginManager, twinkle.strokeColor), getTwinkleFillStyle = () => {
            if (!twinkleFillRgb) {
                return undefined;
            }
            return getStyleFromHsl(twinkleFillRgb, container.hdr, twinklingOpacity);
        }, getTwinkleStrokeStyle = () => {
            if (!twinkleStrokeRgb) {
                return undefined;
            }
            return getStyleFromHsl(twinkleStrokeRgb, container.hdr, twinklingOpacity);
        }, twinkleFillStyle = getTwinkleFillStyle(), twinkleStrokeStyle = getTwinkleStrokeStyle(), res = {}, needsTwinkle = twinkling && (!!twinkleFillStyle || !!twinkleStrokeStyle);
        res.fill = needsTwinkle ? twinkleFillStyle : undefined;
        res.stroke = needsTwinkle ? twinkleStrokeStyle : undefined;
        return res;
    }
    init() {
    }
    isEnabled(particle) {
        const pOptions = particle.options, twinkleOptions = pOptions.twinkle;
        if (!twinkleOptions) {
            return false;
        }
        return twinkleOptions.particles.enable;
    }
    loadOptions(options, ...sources) {
        options.twinkle ??= new Twinkle();
        for (const source of sources) {
            options.twinkle.load(source?.twinkle);
        }
    }
    update() {
    }
}
