// @tsparticles/interaction-external-trail/cjs/Options/Classes/Trail.js
import { deepExtend, isNull, } from "@tsparticles/engine";
import { TrailColorCoords } from "./TrailColorCoords.js";
export class Trail {
    colorCoords;
    delay;
    particles;
    pauseOnStop;
    quantity;
    constructor() {
        this.delay = 1;
        this.pauseOnStop = false;
        this.quantity = 1;
        this.colorCoords = new TrailColorCoords();
    }
    load(data) {
        if (isNull(data))
            return;
        if (data.delay !== undefined) {
            this.delay = data.delay;
        }
        if (data.quantity !== undefined) {
            this.quantity = data.quantity;
        }
        if (data.pauseOnStop !== undefined) {
            this.pauseOnStop = data.pauseOnStop;
        }
        if (data.particles !== undefined) {
            this.particles = deepExtend({}, data.particles);
        }
        if (data.colorCoords) {
            this.colorCoords ??= new TrailColorCoords();
            this.colorCoords.load(data.colorCoords);
        }
    }
}
