// @tsparticles/updater-destroy/cjs/Utils.js
import { AnimatableColor, PixelMode, getRangeValue, identity, isNumber, itemFromSingleOrMultiple, loadParticlesOptions, randomInRangeValue, setRangeValue, } from "@tsparticles/engine";
const defaultOffset = 0, minDestroySize = 0.5, defaultSplitCount = 0, increment = 1, unbreakableTime = 500, minSplitCount = 0, minHue = 0, hueRange = 360, minSaturation = 0, maxSaturation = 100, minLightness = 0, maxLightness = 100;
function createParentColor(parentColor) {
    return AnimatableColor.create(undefined, {
        value: {
            hsl: parentColor,
        },
    });
}
function createOffsetColor(parentColor, offset) {
    const offsetH = getRangeValue(offset.h ?? defaultOffset), offsetS = getRangeValue(offset.s ?? defaultOffset), offsetL = getRangeValue(offset.l ?? defaultOffset), h = (parentColor.h + offsetH) % hueRange, s = Math.max(minSaturation, Math.min(maxSaturation, parentColor.s + offsetS)), l = Math.max(minLightness, Math.min(maxLightness, parentColor.l + offsetL));
    return AnimatableColor.create(undefined, {
        value: {
            hsl: {
                h: h < minHue ? h + hueRange : h,
                s,
                l,
            },
        },
    });
}
function resolveSplitColor(offset, splitColor, splitParticlesColor, parentColor) {
    if (offset && parentColor) {
        return createOffsetColor(parentColor, offset);
    }
    if (splitColor) {
        return AnimatableColor.create(undefined, splitColor);
    }
    if (splitParticlesColor) {
        return AnimatableColor.create(undefined, splitParticlesColor);
    }
    return parentColor ? createParentColor(parentColor) : undefined;
}
function addSplitParticle(pluginManager, container, parent, splitParticlesOptions) {
    const destroyOptions = parent.options.destroy;
    if (!destroyOptions) {
        return;
    }
    const splitOptions = destroyOptions.split, splitParticleOptions = loadParticlesOptions(pluginManager, container, parent.options), splitFillColor = itemFromSingleOrMultiple(splitOptions.fillColor), splitStrokeColor = itemFromSingleOrMultiple(splitOptions.strokeColor), parentFillColor = parent.getFillColor(), parentStrokeColor = parent.getStrokeColor();
    splitParticleOptions.move.load({
        center: {
            x: parent.position.x,
            y: parent.position.y,
            mode: PixelMode.precise,
        },
    });
    const factor = identity / getRangeValue(splitOptions.factor.value), sizeOptions = splitParticleOptions["size"];
    if (sizeOptions) {
        if (isNumber(sizeOptions.value)) {
            sizeOptions.value *= factor;
        }
        else {
            sizeOptions.value.min *= factor;
            sizeOptions.value.max *= factor;
        }
    }
    splitParticleOptions.load(splitParticlesOptions);
    const splitParticlePaintOptions = itemFromSingleOrMultiple(splitParticleOptions.paint), splitParticleFillOptions = splitParticlePaintOptions?.fill, splitParticleStrokeOptions = splitParticlePaintOptions?.stroke, fillColor = resolveSplitColor(splitOptions.fillColorOffset, splitFillColor, splitParticleFillOptions?.color, parentFillColor), strokeColor = resolveSplitColor(splitOptions.strokeColorOffset, splitStrokeColor, splitParticleStrokeOptions?.color, parentStrokeColor);
    if (fillColor && splitParticleFillOptions) {
        splitParticleFillOptions.color = fillColor;
    }
    if (strokeColor && splitParticleStrokeOptions) {
        splitParticleStrokeOptions.color = strokeColor;
    }
    const offset = splitOptions.sizeOffset ? setRangeValue(-parent.size.value, parent.size.value) : defaultOffset, position = {
        x: parent.position.x + randomInRangeValue(offset),
        y: parent.position.y + randomInRangeValue(offset),
    };
    return container.particles.addParticle(position, splitParticleOptions, parent.group, (particle) => {
        if (particle.size.value < minDestroySize) {
            return false;
        }
        particle.velocity.length = randomInRangeValue(setRangeValue(parent.velocity.length, particle.velocity.length));
        particle.splitCount = (parent.splitCount ?? defaultSplitCount) + increment;
        particle.unbreakable = true;
        particle.unbreakableUntil = performance.now() + unbreakableTime;
        return true;
    });
}
export function split(pluginManager, container, particle) {
    const destroyOptions = particle.options.destroy;
    if (!destroyOptions) {
        return;
    }
    const splitOptions = destroyOptions.split;
    if (splitOptions.count >= minSplitCount) {
        if (particle.splitCount === undefined || particle.splitCount > splitOptions.count) {
            return;
        }
        particle.splitCount++;
    }
    const rate = getRangeValue(splitOptions.rate.value), particlesSplitOptions = itemFromSingleOrMultiple(splitOptions.particles);
    for (let i = 0; i < rate; i++) {
        addSplitParticle(pluginManager, container, particle, particlesSplitOptions);
    }
}
