// @tsparticles/interaction-external-drag/cjs/Dragger.js
import { ExternalInteractorBase, } from "@tsparticles/plugin-interactivity";
import { defaultFps, getDistance, isInArray, millisecondsToSeconds, } from "@tsparticles/engine";
import { Drag } from "./Options/Classes/Drag.js";
const clickReleaseDelayMs = 50, defaultMomentumFactor = 0.03, dragMode = "drag", firstSampleIndex = 0, lastSampleOffset = 1, maxMomentumSamples = 5, minSamplesForMomentum = 2, momentumToVelocityFactor = millisecondsToSeconds / defaultFps, zeroValue = 0;
export class Dragger extends ExternalInteractorBase {
    maxDistance = 0;
    #dragStartClickTime;
    #dragStartClickingFalseTime;
    #draggedParticle;
    #grabOffset;
    #lastMousePosition;
    #momentumSamples = [];
    #mouseDownHandled = false;
    #savedVelocity;
    constructor(container) {
        super(container);
    }
    clear(_particle, _delta) {
    }
    init() {
    }
    interact(interactivityData, _delta) {
        const mouse = interactivityData.mouse, mousePos = mouse.position;
        if (mousePos) {
            this.#lastMousePosition = mousePos;
        }
        if (this.#draggedParticle) {
            const now = performance.now();
            if (!mouse.clicking) {
                this.#dragStartClickingFalseTime ??= now;
            }
            else {
                this.#dragStartClickingFalseTime = undefined;
            }
            const hasStableClickEndWithoutClickTime = this.#dragStartClickTime === undefined &&
                this.#dragStartClickingFalseTime !== undefined &&
                now - this.#dragStartClickingFalseTime >= clickReleaseDelayMs;
            if (!mouse.inside || this.#hasClickEnded(mouse) || hasStableClickEndWithoutClickTime) {
                this.#releaseDragged();
                this.#mouseDownHandled = false;
                return;
            }
            const pointerPos = mousePos ?? this.#lastMousePosition;
            if (!pointerPos) {
                return;
            }
            this.#momentumSamples.push({ x: pointerPos.x, y: pointerPos.y, t: now });
            if (this.#momentumSamples.length > maxMomentumSamples) {
                this.#momentumSamples.shift();
            }
            const targetX = pointerPos.x + (this.#grabOffset?.x ?? zeroValue), targetY = pointerPos.y + (this.#grabOffset?.y ?? zeroValue);
            this.#draggedParticle.position.x = targetX - this.#draggedParticle.offset.x;
            this.#draggedParticle.position.y = targetY - this.#draggedParticle.offset.y;
            this.#draggedParticle.velocity.x = zeroValue;
            this.#draggedParticle.velocity.y = zeroValue;
            this.#draggedParticle.initialPosition.x = this.#draggedParticle.position.x;
            this.#draggedParticle.initialPosition.y = this.#draggedParticle.position.y;
            this.#draggedParticle.misplaced = false;
            return;
        }
        if (!mouse.clicking) {
            this.#releaseDragged();
            this.#mouseDownHandled = false;
            return;
        }
        if (this.#mouseDownHandled) {
            return;
        }
        this.#mouseDownHandled = true;
        const downPos = mouse.downPosition ?? mousePos ?? this.#lastMousePosition, closest = this.#findParticleUnderCursor(interactivityData, downPos);
        if (!closest || !mousePos) {
            if (!closest) {
                this.#mouseDownHandled = false;
            }
            return;
        }
        const renderedPos = closest.getPosition();
        this.#draggedParticle = closest;
        this.#dragStartClickTime = mouse.clickTime;
        this.#grabOffset = {
            x: renderedPos.x - mousePos.x,
            y: renderedPos.y - mousePos.y,
        };
        this.#savedVelocity = { x: closest.velocity.x, y: closest.velocity.y };
        this.#dragStartClickingFalseTime = undefined;
        this.#momentumSamples = [{ x: mousePos.x, y: mousePos.y, t: performance.now() }];
        closest.velocity.x = zeroValue;
        closest.velocity.y = zeroValue;
        closest.position.x = renderedPos.x - closest.offset.x;
        closest.position.y = renderedPos.y - closest.offset.y;
        closest.initialPosition.x = closest.position.x;
        closest.initialPosition.y = closest.position.y;
        closest.misplaced = false;
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, mouse = interactivityData.mouse, events = (particle?.interactivity ?? container.actualOptions.interactivity)?.events;
        if (this.#draggedParticle || this.#mouseDownHandled) {
            return true;
        }
        return !!events?.onClick.enable && mouse.clicking && !!mouse.position && isInArray(dragMode, events.onClick.mode);
    }
    loadModeOptions(options, ...sources) {
        options.drag ??= new Drag();
        for (const source of sources) {
            options.drag.load(source?.drag);
        }
    }
    reset(_interactivityData, _particle) {
    }
    #findParticleUnderCursor(interactivityData, mousePos) {
        if (!mousePos) {
            return;
        }
        const candidates = this.container.particles.filter(p => this.isEnabled(interactivityData, p));
        let closest, closestDist = Infinity;
        for (const candidate of candidates) {
            const dist = getDistance(candidate.getPosition(), mousePos), radius = candidate.getRadius();
            if (!Number.isFinite(dist) || !Number.isFinite(radius) || dist > radius || dist >= closestDist) {
                continue;
            }
            closest = candidate;
            closestDist = dist;
        }
        return closest;
    }
    #hasClickEnded(mouse) {
        return (this.#dragStartClickTime !== undefined &&
            mouse.clickTime !== undefined &&
            mouse.clickTime !== this.#dragStartClickTime);
    }
    #releaseDragged() {
        if (this.#draggedParticle) {
            const dragOptions = this.container.actualOptions.interactivity?.modes.drag, preserveMomentum = dragOptions?.preserveMomentum ?? false;
            if (preserveMomentum && this.#momentumSamples.length >= minSamplesForMomentum) {
                const first = this.#momentumSamples[firstSampleIndex], last = this.#momentumSamples[this.#momentumSamples.length - lastSampleOffset];
                if (first && last) {
                    const dt = last.t - first.t;
                    if (dt > zeroValue) {
                        const factor = dragOptions?.momentumFactor ?? defaultMomentumFactor;
                        this.#draggedParticle.velocity.x = ((last.x - first.x) / dt) * momentumToVelocityFactor * factor;
                        this.#draggedParticle.velocity.y = ((last.y - first.y) / dt) * momentumToVelocityFactor * factor;
                    }
                    else if (this.#savedVelocity) {
                        this.#draggedParticle.velocity.x = this.#savedVelocity.x;
                        this.#draggedParticle.velocity.y = this.#savedVelocity.y;
                    }
                }
                else if (this.#savedVelocity) {
                    this.#draggedParticle.velocity.x = this.#savedVelocity.x;
                    this.#draggedParticle.velocity.y = this.#savedVelocity.y;
                }
            }
            else if (this.#savedVelocity) {
                this.#draggedParticle.velocity.x = this.#savedVelocity.x;
                this.#draggedParticle.velocity.y = this.#savedVelocity.y;
            }
        }
        this.#draggedParticle = undefined;
        this.#dragStartClickTime = undefined;
        this.#dragStartClickingFalseTime = undefined;
        this.#grabOffset = undefined;
        this.#lastMousePosition = undefined;
        this.#momentumSamples = [];
        this.#savedVelocity = undefined;
    }
}
